XEN host up / Add host_ip

jonathan-correa-paiva opened the pull request:
To check if an XCP server is up or not, we can serve a metric xen_host_up with the status of the server, for example if we have a [Errno 113] No route to host exception. By using "xen_host_up," you can receive alerts whenever a host is down.
Here's an example:

![image](https://github.com/MikeDombo/xen-exporter/assets/101902399/9b9a4695-8fad-4ad3-8850-4ee69eb95f7b)

Another option is to create a Grafana Panel that provides a quick overview of the status of XCP hosts.
The values are as follows: 0 for Host Down and 1 for Host Up.

![image](https://github.com/MikeDombo/xen-exporter/assets/101902399/b4ddb935-6283-49b3-b7dd-dbef169a2f13)
![image](https://github.com/MikeDombo/xen-exporter/assets/101902399/67898ff5-8e8a-49ff-b1df-dad0996284e1)

With this setup and the host_ip included in the host metrics, we can provide a link to the panel that filters the results based on the host_ip, allowing for quick access.

Any comments and/or suggestions are welcome. this is a Basic concept by now.

MikeDombo commented:
Let me know when this is ready for review by marking the PR as "ready for review".

jonathan-correa-paiva commented:
I didn't have much time to continue on this, this changes are working as spected in my environment since the day I created the PR.

I need to see if the exporter is connected to a pool master, if we need some other chage.

Laclan commented:
Hey Everyone,

I have deployed and tested the changes in my docker environment and confirm that all the features are working.

garrett928 commented:
What's the state of this PR? I can test it tomorrow if another review is what is needed.

jonathan-correa-paiva commented:
This is ready for review, I was thinkin to cash the name of the servers, but maybe it's for another issue.

I'm using this in my production environment scince I've created this pr and it's working as intended.

jonathan-correa-paiva commented:
I had to resolve some conflicts, there would be necessary to review if I didn't break anything, I wouls start testing this changes

jadsy2107 commented:
Sounds good Is there a new JSON dashboard available ? On 8 Aug 2023, at 4:36 am, Jonathan Correa Paiva ***@***.***> wrote:﻿
This is ready for review, I was thinkin to cash the name of the servers, but maybe it's for another issue.
I'm using this in my production environment scince I've created this pr and it's working as intended.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

jonathan-correa-paiva commented:
This is the configuration of the panel I'm using, 

```{
  "id": 17,
  "gridPos": {
    "h": 18,
    "w": 4,
    "x": 6,
    "y": 1
  },
  "type": "stat",
  "title": "XCP Host status",
  "transformations": [],
  "datasource": {
    "uid": "uid",
    "type": "prometheus"
  },
  "pluginVersion": "9.2.8",
  "fieldConfig": {
    "defaults": {
      "mappings": [
        {
          "options": {
            "0": {
              "index": 1,
              "text": "Off"
            },
            "1": {
              "index": 0,
              "text": "On"
            }
          },
          "type": "value"
        }
      ],
      "thresholds": {
        "mode": "absolute",
        "steps": [
          {
            "color": "green",
            "value": null
          },
          {
            "color": "red",
            "value": 0
          },
          {
            "color": "dark-green",
            "value": 1
          }
        ]
      },
      "color": {
        "mode": "thresholds"
      },
      "links": [
        {
          "targetBlank": true,
          "title": "",
          "url": "link"
        }
      ]
    },
    "overrides": []
  },
  "options": {
    "reduceOptions": {
      "values": false,
      "calcs": [
        "lastNotNull"
      ],
      "fields": ""
    },
    "orientation": "horizontal",
    "textMode": "value_and_name",
    "colorMode": "background",
    "graphMode": "area",
    "justifyMode": "auto"
  },
  "targets": [
    {
      "datasource": {
        "type": "prometheus",
        "uid": "uid"
      },
      "editorMode": "code",
      "exemplar": false,
      "expr": "xen_host_up",
      "format": "time_series",
      "hide": false,
      "instant": true,
      "legendFormat": "{{host_ip}} ",
      "range": false,
      "refId": "A"
    }
  ]
}```

I've obfuscated some values, but, it's only using the xen_host_up, witho a couple of Value Mapping and thresholds to show the colors.

Additionally I'm using a link to a custom panel for each xen_host

jadsy2107 commented:
Has it been built into new image ? ghcr.io/mikedombo/xen-exporter:latest

jadsy2107 commented:
> 

That doesn't seem to be complete, are you able to share the latest JSON for Dashboard ?

thanks

jonathan-correa-paiva commented:
I shared a panel that I have on my main dashboard with other values. I'm not using it in the dashboard commented on this repo. Let me create a new one so you can import it.

But for now, this panel is what I shared previously in the image.

MikeDombo commented on xen-exporter.py:
```diff
@@ -139,109 +139,120 @@ def collect_metrics():
+    collector_end_time = time.perf_counter()
+    output += collect_sr_usage(xen)
+    output += f"xen_collector_duration_seconds {collector_end_time - collector_start_time}\n"
+    xen_host_up_extra_tags = {}
```
This block seems quite a bit more complicated than necessary. You just need to output the xen_host_up value and single ip tag. In this case the extra tags dict always only has one value in it.

jonathan-correa-paiva replied:
Simplified as suggested — xen_host_up is now emitted as a single line per host with just the host_ip tag (no extra dict/loop).

MikeDombo reviewed (commented):
Apologies for just getting to this now. I missed some of the updates when you said it was ready to review.

A couple small comments to address if you're able.

Thank you!

jonathan-correa-paiva commented on xen-exporter.py:
```diff
@@ -139,109 +139,120 @@ def collect_metrics():
+        print(e)
+        xen_host_up = False
+    collector_end_time = time.perf_counter()
+    output += collect_sr_usage(xen)
```
Fixed — collect_sr_usage() is now called once, outside the per-host loop, no duplicate call.

jonathan-correa-paiva commented:
Update: pushed the fixes for both review comments above (resolved the threads) and rebased the branch onto the current `master`, which had diverged quite a bit since this was opened (multi-host support via `XEN_MODE`, exception handling fixes, etc.). `xen_host_up` now works per-host inside that loop instead of the old single-host flow.

Changes in the latest push:
- Removed the duplicate `collect_sr_usage(xen)` call.
- Simplified the tagging: `xen_host_up{host_ip="..."}` is now a single output line per host instead of threading the tag through the whole metric loop.
- Each host's metrics collection is wrapped in its own try/except, so one unreachable host reports `xen_host_up 0` and the scrape continues for the rest instead of failing entirely (this also fixes a partial-output edge case I found while rebasing: if a host failed mid-collection, some of its metric lines could leak into the output alongside `xen_host_up 0` — now a host's lines are only kept if it fully succeeds).

Apologies for letting this go quiet for a couple years — happy to keep iterating if there's more to address. This one would genuinely help us right now: we're moving to a policy of only running the published image (`ghcr.io/mikedombo/xen-exporter`) instead of custom builds, and `xen_host_up` is the metric we rely on for host-down alerting, so having it upstream would save us from maintaining a fork just for this.

jonathan-correa-paiva commented:
Quick update on testing: this is now running against 7 real production XCP hosts (single-host mode, `ghcr.io/mikedombo/xen-exporter:latest` + bind-mounted script), all healthy. Sample `/metrics` output (IPs, UUIDs, host/VM names redacted, trimmed to the relevant lines):

```
xen_host_memory_total_kib{host_uuid="aaaaaaaa-0000-0000-0000-000000000001", host="xcp-host-1", host_ip="10.0.0.11"} 33379872.0
xen_host_memory_free_kib{host_uuid="aaaaaaaa-0000-0000-0000-000000000001", host="xcp-host-1", host_ip="10.0.0.11"} 7739824.0
xen_host_cpu_avg{host_uuid="aaaaaaaa-0000-0000-0000-000000000001", host="xcp-host-1", host_ip="10.0.0.11"} 0.5848
xen_host_cpu{host_uuid="aaaaaaaa-0000-0000-0000-000000000001", host="xcp-host-1", host_ip="10.0.0.11", cpu="0"} 0.533
xen_host_loadavg{host_uuid="aaaaaaaa-0000-0000-0000-000000000001", host="xcp-host-1", host_ip="10.0.0.11"} 0.3525
xen_host_pool_session_count{host_uuid="aaaaaaaa-0000-0000-0000-000000000001", host="xcp-host-1", host_ip="10.0.0.11"} 7.0
xen_host_sr_cache_hits{host="xcp-host-1", host_uuid="aaaaaaaa-0000-0000-0000-000000000001", sr_uuid="bbbbbbbb-0000-0000-0000-000000000001", host_ip="10.0.0.11", sr="Local storage"} 0.0
xen_host_iops_total{host="xcp-host-1", host_uuid="aaaaaaaa-0000-0000-0000-000000000001", sr_uuid="bbbbbbbb-0000-0000-0000-000000000001", host_ip="10.0.0.11", sr="Local storage"} 3.0033
xen_vm_cpu{vm_uuid="cccccccc-0000-0000-0000-000000000001", cpu="0", vm="vm-1"} 0.451
xen_vm_memory{vm_uuid="cccccccc-0000-0000-0000-000000000001", vm="vm-1"} 4297052160.0
xen_vm_vbd_write{vm_uuid="cccccccc-0000-0000-0000-000000000001", vm="vm-1", vbd="xvda"} 18239.6816
xen_vm_cpu{cpu="0", vm="Control domain on host: xcp-host-1", vm_uuid="cccccccc-0000-0000-0000-000000000002"} 0.0478
xen_collector_duration_seconds 0.09576406702399254
xen_host_up{host_ip="10.0.0.11"} 1
```

Everything else (all the other `xen_host_*`/`xen_vm_*` lines) came through unaffected, exactly as before this branch. Let me know if there's anything else you'd like changed.

jonathan-correa-paiva commented:
@MikeDombo whenever you get a chance, this should be ready for another look.

MikeDombo reviewed (approved):
LGTM
